Rust support documentation install guide [WIP] (#445)
leaningtech/labs: sites/browserpod/src/content/docs/12-guides/06-installing-rust-toolchain.md

---
title: Installing the Rust toolchain

> description: How to install and use the browserpod Rust toolchain.

Reviewer: ```suggestion
description: How to install and use the BrowserPod Rust toolchain.
```

---

This guide will show you how to install and use the **BrowserPod Rust toolchain** that allows one to compile Rust programs with BrowserPod as the target. Allowing you to run the compiled binaries in the pod.

> Rust is a compiled language and there is no Rust runtime inside a pod, the installation instead is about installing the browserpod toolchain that enables you to compile Rust programs on your _host_ machine, into a binary that can run in browserpod. (also known as [cross compiling](/docs/more/glossary#cross-compilation)).

Reviewer: ```suggestion
Rust is a compiled language, and there is no Rust runtime inside a Pod. Instead, this installation sets up the BrowserPod toolchain on your _host_ machine, which lets you compile Rust programs into a binary that can run inside a Pod (a process known as [cross-compilation](/docs/more/glossary#cross-compilation)).
```

## Installing and setting up the toolchain

### Prerequisites

- [rustup](https://rustup.rs/#)

> ### 1. install the BrowserPod Rust compilation toolchain

Reviewer: ```suggestion
### 1. Install the BrowserPod Rust compilation toolchain
```

The easiest way to install the toolchain is by running our _install script_:

```bash
curl https://rt.browserpod.io/%BP_LATEST%/rust/install.sh | bash
```

> _Alternatively you can download the rarvakk from https://rt.browserpod.io/%BP_LATEST%/rust/browserpod-rust-%BP_LATEST%.tar.gz and follow the steps of the `install.sh` contained within._

Reviewer: ```suggestion
_Alternatively, you can download the tarball from [here](https://rt.browserpod.io/%BP_LATEST%/rust/browserpod-rust-%BP_LATEST%.tar.gz) and follow the steps of the `install.sh` contained within._
```

> This will install a particular Rust nightly version using rustup, and install the toolchain needed to compile Rust to browserpod.

Reviewer: ```suggestion
This will install a particular Rust nightly version using rustup, and install the toolchain needed to compile Rust to BrowserPod.
```

### 2. Set BrowserPod as the active project toolchain

By default, rustup uses your default toolchain (usually `stable`), which doesn't know about the BrowserPod target. Setting an override tells rustup to use the BrowserPod toolchain whenever you build in this directory.

```bash
rustup override set browserpod-%BP_LATEST%
```

> Alternatively you can set `browserpod-%BP_LATEST%` as the active channel in `rust-toolchain.toml`

Reviewer: ```suggestion
Alternatively, you can set `browserpod-%BP_LATEST%` as the active channel in `rust-toolchain.toml`.
```

> _If you'd like to verify the installations success, You can run `rustup show active-toolchain` and see if browserpod is in the outputted list._

Reviewer: ```suggestion
_If you'd like to verify the installation's success, you can run `rustup show active-toolchain` and see if BrowserPod is in the outputted list._
```

### 3. Build your binary

> Now we can build our browserpod wasm binary!

Reviewer: ```suggestion
Now we can build our BrowserPod Wasm binary!
```

```bash
cargo build --release --target wasm32-browserpod-linux-musl
```

> That's it! You now have a binary that can run your Rust program directly in a [Pod](/docs/reference/BrowserPod)! You can find the binary in `target/wasm32-browserpod-linux-musl/release/<binary name>`.

Reviewer: ```suggestion
That's it! You now have a binary that can run directly in a [Pod](/docs/reference/BrowserPod)! You can find the binary in `target/wasm32-browserpod-linux-musl/release/<binary name>`.
```

> To see how to launch a wasm binary in a [Pod](/docs/reference/BrowserPod), see our [Write files to the Pod](/docs/guides/write-files-to-pod) guide!

Reviewer: ```suggestion
To see how to launch a Wasm binary in a [Pod](/docs/reference/BrowserPod), see our [Write files to the Pod](/docs/guides/write-files-to-pod) guide!
```

## Reducing binary size

> Rust builds, by default are both unoptimised and carry debug info. Which, while helpfull for debugging and build speed, can significantly increase browser load times and worst case cause your binary to fail to load all together.

Reviewer: ```suggestion
Rust builds are both unoptimised and carry debug info by default, which while helpful for debugging and build speed, it can significantly increase browser load times. In the worst case, it will cause your binary to fail to load altogether.
```

Author: think the it should be removed here. But I restructured the sentence entirely instead.

"Rust builds carry debug info and remain unoptimised by default. Although this aids debugging and speeds up compilation, it inflates browser load times and, in the worst case, prevents the binary from loading at all."

This is why we add the `--release` flag when building. To reduce the build size further consider creating a custom build profile in your `Cargo.toml` and setting the following settings.

> | settings          | effect                                                     | cost                         |

Reviewer: ```suggestion
| Settings          | Effect                                                     | Cost                         |
```

| ----------------- | ---------------------------------------------------------- | ---------------------------- |

> | lto = true        | Removes unused code across crate boundaries                | longer build time            |

Reviewer: ```suggestion
| `lto = true`        | Removes unused code across crate boundaries                | Longer build time            |
```

> | codegen-units = 1 | Combine all code in a crate into a single compilation unit | longer build time            |

Reviewer: ```suggestion
| `codegen-units = 1 `| Combine all code in a crate into a single compilation unit | Longer build time            |
```

> | opt-level = "z"   | Optimize binary size over runtime speed                    | Worse program performance    |

Reviewer: ```suggestion
| `opt-level = "z"`   | Optimize binary size over runtime speed                    | Worse program performance    |
```

> | strip = true      | Removes debug symbols                                      | Less useful panic backtraces |

Reviewer: ```suggestion
| `strip = true`      | Removes debug symbols                                      | Less useful panic backtraces |
```

A custom profile with all settings enabled would look like:

```toml
[profile.release-small]
inherits = "release"
lto = true
codegen-units = 1
strip = true
```

Then to build with said profile, use the `--profile` flag.

```bash
cargo build --profile release-small --target wasm32-browserpod-linux-musl
```

_(Note that cargo creates a different build directory per profile, so the created build will now reside in: `target/wasm32-browserpod-linux-musl/release-small/<binary name>`.)_

## Uninstalling and/or updating

> If you'd like to uninstall the toolchain, you can run first run `rustup toolchain list` to find what version you have, and then run:

Reviewer: ```suggestion
If you'd like to uninstall the toolchain, you can first run `rustup toolchain list` to find what version you have, and then run:
```

```bash
rustup toolchain uninstall browserpod-<your browserpod version>
```

For updating, first uninstall, and then reinstall the updated version.

## Tips and Troubleshooting

> ### If your project has C dependencies

Reviewer: I think this goes better under the pre requisites section

You will need `clang` and `llvm` as added prerequisites. You can check

> `clang --print-targets | grep -i wasm` and `which llvm-ar` to check for their presence respectively.

Reviewer: ```suggestion
You will need `clang` and `llvm` as added prerequisites. You can check `clang --print-targets | grep -i Wasm` to verify your clang build supports Wasm. Also use the command `which llvm-ar` to verify llvm is installed.
```

### Build fails with "unable to create target"

If you encounter:

```bash
error: unable to create target: 'No available targets are
compatible with triple "wasm32-unknown-unknown"'
```

> Your clang was built without a WebAssembly backend. Most common on macOS,

Reviewer: ```suggestion
Your clang was built without a WebAssembly backend, most common on macOS
```

---
title: Installing the Rust toolchain
description: How to install and use the browserpod Rust toolchain.
---

>  This guide will show you how to install and use the **BrowserPod Rust toolchain** that allows one to compile Rust programs with BrowserPod as the target. Allowing you to run the compiled binaries in the pod.

Reviewer: ```suggestion
 This guide will show you how to install and use the **BrowserPod Rust toolchain** that allows one to compile Rust programs with BrowserPod as the target. This will enable you to run the compiled binaries in the Pod.
```

Rust is a compiled language and there is no Rust runtime inside a pod, the installation instead is about installing the browserpod toolchain that enables you to compile Rust programs on your _host_ machine, into a binary that can run in browserpod. (also known as [cross compiling](/docs/more/glossary#cross-compilation)).

## Installing and setting up the toolchain

### Prerequisites

> - [rustup](https://rustup.rs/#)

Reviewer: ```suggestion
- [rustup](https://rustup.rs/) installed on your machine
```

---
title: Installing the Rust toolchain
description: This tutorial will guide you on installing the browserpod rust toolchain, that allows you to compile rust programs into a binary that can run in browserpod
---

> In this tutorial you will be guided in installing the **Browserpod rust toolchain** that allows you to compile rust programs with browserpod as the target (also known as _crosscompiling_). Allowing you to run the compiled binaries in the pod.

Reviewer: ```suggestion
In this tutorial you will be guided in installing the **BrowserPod rust toolchain** that allows one to compile rust programs with BrowserPod as the target (also known as _crosscompiling_). Allowing you to run the compiled binaries in the pod.
```

> Because Rust is a compiled language and there is no rust runtime inside a pod, the installation instead is about installing the browserpod toolchain that enables you to compile rust programs on your _host_ machine, into a binary that can run in browserpod.

Reviewer: Please start with proper paragraph wording here, starting a paragraph with "Because" sounds odd.

You can start with "Rust is a compiled language that..."

_!!! Alt: less mention of host machine and/or more about cross compilation and different targets (or different wording)_
!!! or more explecit mention of no rust runtime being present inside the pods

> ### prerequisites needed

Reviewer: ```suggestion
### Prerequisites
```

> ### prerequisites needed

Reviewer: Please also move this under "## Installing and setting up the toolchain"

- [rustup](https://rustup.rs/#)

> !!! Could add clang and llvm if C dependencies need to be added (with potential install steps or command of checking if said prerequisites are present on host machine) but not sure if overkill.

Reviewer: Are these exclamation marks across the whole document comments?

Author: yes! I mentioned it to Alessandro in person but forgot to mention it in the PR. What do you think of the format?

Either way for this stage I've moved the comments into my PR message

> # Installing the toolchain

Reviewer: ```suggestion
## Installing and setting up the toolchain
```

> ## 1. install the browserpod rust compilation toolchain

Reviewer: ```suggestion
## 1. Install the BrowserPod rust compilation toolchain
```

> ## 1. install the browserpod rust compilation toolchain

Reviewer: ```suggestion
### 1. install the BrowserPod rust compilation toolchain
```

The easiest way to install the toolchain is by running our _install script_:

```bash
curl https://rt.browserpod.io/%BP_LATEST%/rust/install.sh | bash
```

_Alternatively you can download the rarvakk from https://rt.browserpod.io/%BP_LATEST%/rust/browserpod-rust-%BP_LATEST%.tar.gz and follow the steps of the `install.sh` contained within._

This will install a particular rust nightly version using rustup, and install the toolchain needed to compile rust to browserpod.

> ## 2. Set browserpod as the active project toolchain

Reviewer: ```suggestion
## 2. Set BrowserPod as the active project toolchain
```

> ## 2. Set browserpod as the active project toolchain

Reviewer: ```suggestion
### 2. Set BrowserPod as the active project toolchain
```

Now that we have our toolchain installed, we can head into any of our rust projects and set browserpod as its active toolchain.

By default, rustup uses your default toolchain (usually `stable`), which doesn't know about the BrowserPod target. Setting an override tells rustup to use the BrowserPod toolchain whenever you build in this directory.

!! Can add a cd line project line and a potential cargo --hello world command line for those not having rust projects yet.

```bash
rustup override set browserpod-%BP_LATEST%
```

Alternatively you can set `browserpod-%BP_LATEST%` as the active channel in `rust-toolchain.toml`

_If you'd like to verify the installations success, You can run `rustup show active-toolchain` and see if browserpod is in the outputted list._

> ## 3. Build your binary

Reviewer: ```suggestion
### 3. Build your binary
```

Now we can build our browserpod wasm binary!

```bash
cargo build --release --target wasm32-browserpod-linux-musl
```

> that's it. You now run your rust program directly in the browser! You can find the binary in `target/wasm32-browserpod-linux-musl/release/<binary name>`.

Reviewer: ```suggestion
That's it. You now run your rust program directly in the browser! You can find the binary in `target/wasm32-browserpod-linux-musl/release/<binary name>`.
```

> To see how to launch a wasm binary in a [Pod](). see our [loading binaries in the pod](TBA) guide!

Reviewer: ```suggestion
To see how to launch a wasm binary in a Pod. see our [loading binaries in the pod](TBA) guide!
```

Author: Ah I thought a link to the term in the glossary would be a nice small UX improvement. Do you think it's too much?

(also question if that warrants a full guide but might be useful for both rust and less technical rust users? also for future)

> ## uninstalling and/or updating

Reviewer: ```suggestion
## Uninstalling and/or updating
```
